src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+    std::deque<CompressedHeader>().swap(m_redownloaded_headers);
+    m_redownload_buffer_last_hash.SetNull();
+    m_redownload_buffer_first_prev_hash.SetNull();
+    m_process_all_remaining_headers = false;
+    m_current_height = 0;
+    std::vector<uint256>().swap(m_chain_start_locator.vHave);
+
+    m_download_state = State::FINAL;
+}
+
+/** Initialize the parameters for this headers download, validate this first
+ * batch, and request more headers. */
+std::optional<CBlockLocator> HeadersSyncState::StartInitialDownload(const CBlockIndex* chain_start,
+        const std::vector<CBlockHeader>& initial_headers, const arith_uint256&
+        minimum_required_work, CBlockLocator&& chain_start_locator)
+{
+    // A new instance of this object should be instantiated for every headers
+    // sync, so that we don't reuse our salted hasher between syncs.
+    assert(m_download_state == State::UNSTARTED);
+    m_download_state = State::INITIAL_DOWNLOAD;
+
+    m_chain_start = chain_start;
+    m_minimum_required_work = minimum_required_work;
+    m_current_chain_work = chain_start->nChainWork;
+    m_current_height = chain_start->nHeight;
+    m_chain_start_locator = std::move(chain_start_locator);
+
+    m_last_header_received = m_chain_start->GetBlockHeader();
+
+    // Estimate the number of blocks that could possibly exist on the peer's
+    // chain *right now* using 6 blocks/second (fastest blockrate given the MTP
+    // rule) times the number of seconds from the last allowed block until
+    // today. This serves as a memory bound on how many commitments we might
+    // store from this peer, and we can safely give up syncing if the peer
+    // exceeds this bound, because it's not possible for a consensus-valid
+    // chain to be longer than this (at the current time -- in the future we
+    // could try again, if necessary, to sync a longer chain).
+    m_max_commitments = 6*(GetAdjustedTime() - chain_start->GetMedianTimePast() + MAX_FUTURE_BLOCK_TIME) / HEADER_COMMITMENT_FREQUENCY;
+
+    if (!ValidateAndStoreHeadersCommitments(initial_headers)) {
+        return std::nullopt;
+    }
+    return MakeNextHeadersRequest();
+}
+
+/** Process the next batch of headers received from our peer.
+ *  Validate and store commitments, and compare total chainwork to our target to
+ *  see if we can switch to REDOWNLOAD mode.  */
+HeadersSyncState::ProcessingResult HeadersSyncState::ProcessNextHeaders(const
+        std::vector<CBlockHeader>& headers, const bool full_headers_message)
+{
+    ProcessingResult ret;
+
+    Assume(!headers.empty());
+    if (headers.empty()) return ret;
+
+    Assume(m_download_state != State::FINAL);
+    if (m_download_state == State::FINAL) return ret;
+
+    if (m_download_state == State::INITIAL_DOWNLOAD) {
+        // During INITIAL_DOWNLOAD, we minimally validate block headers and
+        // occasionally add commitments to them, until we reach our work
+        // threshold.
```

Comment:
5691c174c5cb2f9947e6cc2db8d32b4b4715e47d: maybe add a comment "m_download_state is updated to REDOWNLOAD if headers reach the minimum difficulty."

Reply:
Done.